src/main/java/com/iota/iri/hash/Kerl.java

```diff
@@ -85,63 +91,108 @@ public void squeeze(final int[] trits, int offset, int length) {
     public static BigInteger bigIntFromTrits(final int[] trits, final int offset, final int size) {
 
         BigInteger value = BigInteger.ZERO;
-
-        for (int i = size; i-- > 0; ) {
-
-            value = value.multiply(BigInteger.valueOf(Converter.RADIX)).add(BigInteger.valueOf(trits[offset + i]));
+        int power = 0;
+        for (int i = offset + size - 1; i >= offset; i--) {
+            switch (trits[i]) {
+                case 1: {
+                    int sequence = 0;
+                    while (i - 1 >= offset && trits[i - 1] == 1) {
+                        i--;
+                        sequence++;
+                    }
+                    value = value.multiply(RADIX_POWERS[power + 1 + sequence]).add(RADIX_POSITIVE_ONE_FUNCTION[sequence]);
+                    power = 0;
+                }
+                break;
+                case -1: {
+                    int sequence = 0;
+                    while (i - 1 >= offset && trits[i - 1] == -1) {
+                        i--;
+                        sequence++;
+                    }
+                    value = value.multiply(RADIX_POWERS[power + 1 + sequence]).add(RADIX_NEGATIVE_ONE_FUNCTION[sequence]);
+                    power = 0;
+                }
+                break;
+                default:
+                    power++;
```

Comment:
If it is a ZERO trit, we increment the power so that the next round the power of the multiply is fetched from a lookup.
At the end, if all we have is ZEROS, then the POWER will need to be applied after the loop finishes.